### PR TITLE
DT-2074: add redis, save events to redis and retrieve all events on application startup

### PR DESCRIPTION
**Pull request by petergphillips**

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -11,13 +11,18 @@ jobs:
       localstack_tag: "0.12.9.1"
     steps:
       - checkout
+      - hmpps/install_aws_cli
```
Comment (petergphillips) [inline .circleci/config.yml:14]: copied from probation offender events

File: docker-compose.yml [hunk 2]
```diff
@@ -39,7 +47,7 @@ services:
     volumes:
       - "${TMPDIR:-/tmp/localstack}:/tmp/localstack"
       - "/var/run/docker.sock:/var/run/docker.sock"
-      - $PWD/localstack:/docker-entrypoint-initaws.d
+      - $PWD/src/test/resources/localstack:/docker-entrypoint-initaws.d
```
Comment (petergphillips) [inline docker-compose.yml:50]: changed to be standard with other projects

File: src/main/kotlin/uk/gov/justice/hmpps/offenderevents/App.kt [hunk 3]
```diff
@@ -2,10 +2,22 @@ package uk.gov.justice.hmpps.offenderevents
...
 fun main(args: Array<String>) {
   runApplication<Application>(*args)
 }
+
+// we want keyspace notifications, but have to empty the config parameter (default Ex) since elasticache doesn't support
+// changing the config.  If we move off elasticache then need to remove the config parameter and let it use the default.
+@Configuration
+@EnableRedisRepositories(
+  enableKeyspaceEvents = EnableKeyspaceEvents.ON_STARTUP,
+  keyspaceNotificationsConfigParameter = "\${application.keyspace-notifications:}"
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/hmpps/offenderevents/App.kt:21]: nabbed from token verification api

File: src/main/kotlin/uk/gov/justice/hmpps/offenderevents/config/JmsConfig.kt [hunk 4]
```diff
@@ -89,21 +87,9 @@ class JmsConfig {
   @Bean
   @ConditionalOnProperty(name = ["sqs.provider"], havingValue = "localstack")
   @Suppress("SpringJavaInjectionPointsAutowiringInspection")
-  open fun queueUrl(
+  fun queueUrl(
     @Autowired awsSqsClient: AmazonSQS,
     @Value("\${sqs.queue.name}") queueName: String,
     @Value("\${sqs.dlq.name}") dlqName: String
-  ): String {
-    val result = awsSqsClient.createQueue(CreateQueueRequest(dlqName))
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/hmpps/offenderevents/config/JmsConfig.kt:97]: don't use embedded localstack so don't need to create queues in the kotlin any more

File: src/main/kotlin/uk/gov/justice/hmpps/offenderevents/data/EventRepository.kt [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+import org.springframework.stereotype.Repository
+
+@Repository
+interface EventRepository : CrudRepository<Event, String>
+
+@RedisHash(value = "event", timeToLive = 86400) // expire all tokens after a day
+data class Event(
+  @Id
+  val messageId: String,
+  val wholeMessage: String,
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/hmpps/offenderevents/data/EventRepository.kt:15]: might need to have a ponder on this, but at present dump the whole aws message into redis

File: src/main/kotlin/uk/gov/justice/hmpps/offenderevents/service/OffenderEventStartUp.kt [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+class OffenderEventStartUp(
+  val offenderEventStore: OffenderEventStore,
+  val eventRepository: EventRepository,
+  val gson: com.google.gson.Gson,
+) {
+  companion object {
+    val log: Logger = LoggerFactory.getLogger(this::class.java)
+  }
+
+  @EventListener(ContextRefreshedEvent::class)
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/hmpps/offenderevents/service/OffenderEventStartUp.kt:20]: listener fired on application startup

File: src/main/resources/application-local-redis.yml [hunk 7]
```diff
@@ -0,0 +1,3 @@
+application:
+  # enable keyspace notifications when running in standalone redis (unsupported for elasticache)
+  keyspace-notifications: 'Ex'
```
Comment (petergphillips) [inline src/main/resources/application-local-redis.yml:3]: nabbed from token verification api

File: src/main/kotlin/uk/gov/justice/hmpps/offenderevents/data/EventRepository.kt (hunk 5 above)
Comment (andymarke) [reply src/main/kotlin/uk/gov/justice/hmpps/offenderevents/data/EventRepository.kt:15]: Feels fine to me

**Review by andymarke — APPROVED** [review]
LGTM

**Pull request merged**